src/runtime.cc

```diff
@@ -105,233 +105,323 @@ namespace detail {
-		if(m_test_mode) {
-			assert(m_test_active && "initializing the runtime from a test without a runtime_fixture");
+	runtime::runtime(int* argc, char** argv[], const devices_or_selector& user_devices_or_selector) {
+		m_application_thread = std::this_thread::get_id();
+
+		if(s_test_mode) {
+			assert(s_test_active && "initializing the runtime from a test without a runtime_fixture");
+			s_test_runtime_was_instantiated = true;
 		} else {
 			mpi_initialize_once(argc, argv);
 		}
 
-		int world_size;
-		MPI_Comm_size(MPI_COMM_WORLD, &world_size);
-		m_num_nodes = world_size;
-
 		m_cfg = std::make_unique<config>(argc, argv);
-		if(m_cfg->is_dry_run()) { m_num_nodes = m_cfg->get_dry_run_nodes(); }
 
-		int world_rank;
-		MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
-		m_local_nid = world_rank;
+		if(m_cfg->is_dry_run()) {
+			m_num_nodes = static_cast<size_t>(m_cfg->get_dry_run_nodes());
+			m_local_nid = 0;
+		} else {
+			int world_size = -1;
+			MPI_Comm_size(MPI_COMM_WORLD, &world_size);
+			m_num_nodes = static_cast<size_t>(world_size);
+
+			int world_rank = -1;
+			MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
+			m_local_nid = static_cast<node_id>(world_rank);
+		}
 
-		if(!m_test_mode) { // do not touch logger settings in tests, where the full (trace) logs are captured
+		if(!s_test_mode) { // do not touch logger settings in tests, where the full (trace) logs are captured
 			spdlog::set_level(m_cfg->get_log_level());
-			spdlog::set_pattern(fmt::format("[%Y-%m-%d %H:%M:%S.%e] [{:0{}}] [%^%l%$] %v", world_rank, int(ceil(log10(world_size)))));
+			// TODO is the runtime ctor really the right place to set these globals?
+			spdlog::set_pattern(fmt::format("[%Y-%m-%d %H:%M:%S.%e] [{:0{}}] [%^%l%$] %v", m_local_nid, int(ceil(log10(double(m_num_nodes))))));
 		}
 
+		CELERITY_INFO("Celerity runtime version {} running on {}. PID = {}, build type = {}, {}", get_version_string(), get_sycl_version(), get_pid(),
+		    get_build_type(), get_mimalloc_string());
+
 #ifndef __APPLE__
 		if(const uint32_t cores = affinity_cores_available(); cores < min_cores_needed) {
 			CELERITY_WARN("Celerity has detected that only {} logical cores are available to this process. It is recommended to assign at least {} "
 			              "logical cores. Performance may be negatively impacted.",
 			    cores, min_cores_needed);
 		}
 #endif
-		m_user_bench = std::make_unique<experimental::bench::detail::user_benchmarker>(*m_cfg, static_cast<node_id>(world_rank));
+
+		m_user_bench = std::make_unique<experimental::bench::detail::user_benchmarker>(*m_cfg, m_local_nid);
 
 		cgf_diagnostics::make_available();
 
-		m_h_queue = std::make_unique<host_queue>();
-		m_d_queue = std::make_unique<device_queue>();
+		auto devices = std::visit(
+		    [&](const auto& value) { return pick_devices(m_cfg->get_host_config(), value, sycl::platform::get_platforms()); }, user_devices_or_selector);
+		assert(!devices.empty());
```

Comment:
I think this shouldn't be an assertion but an error check with reporting to the user, since it could be a somewhat common setup error.

Reply:
This exception is already thrown in in `pick_devices`, so the assertion merely checks a postcondition.